src/main/java/ikura/Database.java

```diff
@@ -0,0 +1,242 @@
+        if (xs.isEmpty()) {
+            return new ArrayList<Task>();
+        }
+
+        // TODO: make and use a stream zipper to make this code less obtuse
+        // one might note the flagrant use of Collectors.toList() here, it's because java
+        // streams suck.
+
+        var ret = xs.subList(1, xs.size())
+            .stream()
+            // java is a bad language.
+            .map(new Function<String, Either<String, Task>>() {
+
+                @Override
+                public Either<String, Task> apply(String line) {
+                    /*
+                        format:
+                        <type> <doneness> [date string] "|" <description>
+                        <type>       :: char (ascii char)
+                                        'T' = todo
+                                        'D' = deadline
+                                        'E' = event
+
+                        <doneness>   :: char (ascii char)
+                                        '1' = done
+                                        '0' = rare
+                    */
+
+                    var type = line.charAt(0);
+                    var asdf = line.charAt(1);
+
+                    var done = false;
+
+                    // time for some 3Head java
+                    if (asdf == '1') {
+                        done = true;
+                    } else if (asdf == '0') {
+                        done = false;
+                    } else {
+                        // ideally we'd tell the user the line number, but given that:
+                        // 1. java streams suck, so there's no way to get the index of this line
+                        // 2. they're not supposed to edit the file anyway,
+                        // let's just not bother.
+                        return Either.left(String.format("invalid doneness '%c' (expected either '0' or '1')", asdf));
+                    }
+
+                    Task task = null;
+                    if (type == 'T') {
+
+                        var desc = line.substring(line.indexOf('|') + 1);
+                        task = new Todo(desc);
+
+                    } else if (type == 'D' || type == 'E') {
+
+                        try {
+                            var date = DatedTask.parseDate(line.substring(2, line.indexOf('|')));
+                            var desc = line.substring(line.indexOf('|') + 1);
+
+                            if (type == 'D') {
+                                task = new Deadline(desc, date);
+                            } else {
+                                task = new Event(desc, date);
+                            }
+                        } catch (InvalidInputException e) {
+                            return Either.left(e.toString());
+                        }
+                    } else {
+                        return Either.left(String.format("invalid type '%c' (expected one of 'T', 'D', 'E')", type));
+                    }
```

Comment:
Perhaps there could be less deep nesting to be consistent with the code quality standard of "Avoid Deep Nesting"

Reply:
you're right, i should refactor this method eventually.